### ISSUE TITLE
Support for NCE-loss/sampled softmax with dynamic_decode()

### ISSUE THREAD

**Issue by iamgroot42** [tensorflow/tensorflow#21306]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: NA
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.8.0
- **Python version**: 3.5
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: CUDA 9.1
- **GPU model and memory**: TITAN Xp 12196MB * 8
- **Exact command to reproduce**: NA

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
NCE loss/Sampled Softmax loss both require inputs along with the softmax-weight (used for multiplication to obtain logits). However, the current [sequence_loss()](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/sequence_loss) function in [dynamic_decode()](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode) function expects a loss function that works with logits, not these inputs. To provide these inputs to this custom loss function, the only method right now is to override the [cell ](https://github.com/tensorflow/tensorflow/blob/r1.9/tensorflow/contrib/seq2seq/python/ops/basic_decoder.py#L137) being used, [return](https://github.com/tensorflow/tensorflow/blob/r1.9/tensorflow/contrib/seq2seq/python/ops/decoder.py#L265) their outputs (before passing through output layer) all the way up via its BasicDecoder. Then, this output can be used in the custom loss function.

There should be some support for an optional flag which lets one return the output state (before being passed through that matrix) as well, so that it may be passed to the custom loss function. TO compliment this, the custom loss function should also accept a loss in the format (inputs, outputs) instead if just (logits, outputs).

I spent nearly a week into looking for a workaround for this and could not find anywhere, hence this issue/feature request.

P.S.: If anyone has used NCE loss/sampled softmax loss with dynamic_decode() without going through all of this hassle, please let me know. There is nothing on Stack Overflow/Github/Tensorflow's examples about this (there is one example, but they used the logits as inputs to the sequence_loss's custom loss, which is wrong)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
@lmthang wdyt?

**Comment by lmthang** [comment]
The code has been updated. Check the loss here https://github.com/tensorflow/nmt/blob/master/nmt/model.py#L614 for sampled softmax

**Comment by iamgroot42** [comment]
This works! Thank you so much 😄
